src/main/java/de/aservo/atlassian/confapi/rest/AbstractDirectoriesResourceImpl.java

```diff
@@ -26,31 +25,22 @@ public AbstractDirectoriesResourceImpl(DirectoryService directoryService) {
 
     @Override
     public Response getDirectories() {
-        final ErrorCollection errorCollection = new ErrorCollection();
-        try {
-            final DirectoriesBean directoriesBean = new DirectoriesBean(directoryService.getDirectories());
-            return Response.ok(directoriesBean).build();
-        } catch (Exception e) {
-            log.error(e.getMessage(), e);
-            errorCollection.addErrorMessage(e.getMessage());
-        }
-        return Response.status(BAD_REQUEST).entity(errorCollection).build();
+        final DirectoriesBean directoriesBean = new DirectoriesBean(directoryService.getDirectories());
+        return Response.ok(directoriesBean).build();
```

Comment:
warum gibt es denn hier keinen Fehlerfall mehr?

Reply:
Ich habe für diese Methode keinen Fehlerfall gesehen. Klar, eine `RuntimeException` könnte irgendwie immer fliegen, aber es gibt hier keine erwartbare Exception.

Reply:
Nun ja, wir wissen ja nicht, wie die finale Implementierung dann im jeweiligen Plugin aussieht. Kann ja sein, dass da ein Fehlerfall konstruiert werden soll, der eine 500 o.ä. liefern soll. Die commons sollte dahingehend so generisch sein, dass das möglich ist. 

apropros `RuntimeException` Wie werden die denn eigentlich behandelt? Bruchen wir dafür noch einen speziellen Handler, die die Inhalte ebenfalls in `ErrorCollection` packt?

Reply:
Okay, und wie machen wir das mit den Swagger Annotations? Immer alle / immer keine?

`RuntimeException`, ebenfalls gute Frage. Was denkst du?

Reply:
was hälst du von immer alle? 200, 400, 500. Wenn sie dann tatsächlich per Impl nicht kommen auch gut. Aber so müssten wir nicht nochmal an die commons ran.

Reply:
zu den `RuntimeException`s: aus diesem Grund hatte ich ursprünglich `} catch (Exception e) {` reingebracht, damit fangen wir auch `RuntimeException` ein. Filtern wir jetzt nur noch die checked exceptions (aktueller Stand) dann knallt es hier ggf. zur Laufzeit. Wahrscheinlich gibt tomcat dann auch eine 500 zurück. Auf KEINEN Fall bekommen wir dann eine `ErrorCollection` mehr, was wiederum für den Client nicht so ideal ist.

Reply:
denkbar wäre noch sowas wie (mit dummy method names)
```
} catch (AbstractWebException e) {
      return e.ResponseSpecific();
} catch (Exception e) {
      return e.ResponseAllOther();
}
```

Reply:
Ich habe für diese beiden Vorschläge die Issues #30 und #31 erstellt.

Für das Thema der allgemeinen Error Codes würde ich gerne auch nochmal @brettaufheber befragen, weil er sich gut mit REST API's und Swagger auskennt. Ich hätte aber nichts gegen den Vorschlag einzuwenden.

Über ein allgemeines `catch (Exception e)` bzw. `catch (RuntimeException e)` zum fangen von unchecked Exceptions bin ich aber noch nicht so glücklich. Ich habe aber die Vermutung, dass man diese Art von Exceptions vielleicht sogar in den `ResourceFilter`'s bzw. `ResourceInterceptor`'s behandeln kann. Das würde ich gerne ausprobieren, bevor wir überall die unchecked Exceptions fangen.

Die Issues sind dem Meilenstein für das erste Release zugewiesen, sodass hier erstmal gemergt und weitergemacht werden kann.

Reply:
Prinzipiell sollten immer nur die Exceptions gefangen werden, die man auch erwartet. Für Status-Codes 400 bis 499 ist das Behandeln von Fehlern, wie es hier gemacht wird sinnvoll. Für den Status-Code 500 ist eine Behandlung meiner Meinung nach nicht sinnvoll. Das Web-Framework wird all Exceptions einfangen müssen, um im Fehlerfall nicht von der JVM beendet zu werden. Typisches Verhalten hier ist das schreiben des Stacktrace in eine Log-Datei und die ausgabe des 500er Codes. Bei dem 500er Code sollte es sich nur um Fehler handeln, die nicht dem Web-Kontext zuzuordnen sind. Die Nachrichten dieser Exceptions werden auf gar keinen Fall an den Clienten geleitet, sondern lediglich in eine Log-Datei geschrieben. Bei RESTful-APIs habe ich oft gesehen, dass bei einem 500er Code kein Body mit der Response übertragen wird. Die Logik der Sache ist, dass es sich um einen unerwartete Fehler handelt, weshalb der Server daher gar keine vernünftige Nachricht zur Ausgabe bringen kann.

Reply:
Die stacktraces sollten wir natürlich nicht dem client zurückgeben. Aber ein einheitliches Responsemodell im Fehlerfall ist schon sehr wichtig für die durchgängige Verarbeitbarkeit der REST Responses, egal ob da 400er oder 500er kommen. Ansonsten muss ja der client wieder unterscheiden, ob eine `ErrorCollection` dranhängt oder nicht. 

SpringBoot bsp. erzeugt immer eine Response Struktur im Fehlerfall, für 500er bsp. so
```
{
  "timestamp": 1512714275589,
  "status": 500,
  "error": "Internal Server Error",
  "message": "id-9",
  "path": "/items/2"
}
```

Unser `ErrorCollection` Modell ist momentan noch sehr ein einfach aufgebaut, in der message kann im Zweifel dann auch einfach nur "Internal Server Error" stehen, bei den 400ern entsprechend mehr. Aber immer eine geordnete Struktur und plötzlich ohne ResponseBody.

Reply:
Die meisten Frameworks bieten es an, einen Error-Handler (oder wie auch immer das bei Spring-Boot heißt) zu überschreiben, um Fehler alternativ zu verarbeiten. Für eine einheitliche Formatierung könnte man diesen Mechanismus nutzen. Es gibt sowieso Fehler, an die man ohne den Error-Handler sowieso nicht herankommt, wie zum Beispiel 414: "URI Too Long" oder 431: "Request Header Fields Too Large".

Die Standard-Response vom SpringBoot gefällt mir übrigens auch nicht. Ich stimme zu, dass Einheitlichkeit strebenswert ist. Das der Pfad in so einer Error-Response übertragen wird habe ich schon oft gesehen, finde ich aber unsinnig. Mir fällt auch nach sehr langem Überlegen nicht ein, wozu das gut sein soll.

Meine Stimme hat hier wahrscheinlich nicht so ein hohes Gewicht. Aber dennoch möchte ich mich gegen die Behandlung des 500er-Codes am Famework vorbei aussprechen.

Reply:
Hallo @brettaufheber

danke für den guten Denkanstoß! Ich habe dazu diesen Artikel gefunden, der zeigt, dass sowohl dieses allgemeine Exception Gefange als auch meine eigenen "AbstractWebEx

Reply:
Und gerade noch das hier gefunden, siehe "Default Response". Ich baue mal um...

https://swagger.io/docs/specification/describing-responses/

Reply:
Ok, ich stimme euch zu. Da der Container ja tatsächlich auch selber Exceptions werfen kann und wir Confluence bzgl Responseformate nicht anpassen werden, ist die Vereinheitlichung nutzlos. @pathob Du hast da wirklich einen guten Artikel gefunden. Lass es uns so machen, mit bsp `NotFoundException`, das vereinfacht die Sache erheblich. Ich warte mal auf deinen Umbau. Ich bin dann heute ready for review ;-)